### ISSUE TITLE
Timeout on PurgeOrchestrationInstanceHistoryAsync

### ISSUE THREAD

**Issue by lidong-zhao** [microsoft/durabletask-mssql#137]
It seems mssql provider uses default timeout of 30 seconds for PurgeOrchestrationInstanceHistoryAsync, how can we overwrite the default timeout value?

**Comment by cgillum** [comment]
Are you able to work around it by increasing the connection timeout in the SQL connection string that you're using? It's obviously not ideal, but it seems these APIs need to be updated to support longer timeouts.

**Comment by lidong-zhao** [comment]
Thanks, I could set command timeout in connection string, but that will affect all executions, can we add this support to these APIs?

**Comment by cgillum** [comment]
Yes, we can investigate whether it's possible to add a mechanism that allows you to control the timeout independent from the connection string timeout.

**Comment by lidong-zhao** [comment]
Other than the timeout, another option is to have transaction for each instance Id instead of having one big transaction for all isntance ids.

**Comment by cgillum** [comment]
Oh, I didn't realize we were using a single transaction for all the deletes! Yes, we had to make a similar fix in the Azure Storage provider for Durable Tasks. We should do something similar here. It probably won't be a transaction for each instance ID, but we can at least break it down into smaller batches so that you can at least complete the work by retrying until it's done.

**Comment by lidong-zhao** [comment]
Exactly! Thanks.

**Comment by cgillum** [comment]
@lidong-zhao these operations should be using the clustered indexes, which are defined implicitly. For example, the `NewEvents` table defines the following primary key constraint, which includes both `TaskHub` and `InstanceID`:

```sql
CONSTRAINT PK_NewEvents PRIMARY KEY (TaskHub, InstanceID, SequenceNumber),
```

And here is the estimated query plan from SQL Server:

![image](https://user-images.githubusercontent.com/2704139/211644456-f2ef7390-6e4d-4995-8a9d-33cccd600a77.png)

**Comment by lidong-zhao** [comment]
AH, most of them do have index on taskhub and instance, my mistake that I only checked dt.NewTasks and assumed the rest tables would be the same.
